File: 2024/zig/day-01/main.zig
const std = @import("std");
const print = std.debug.print;

fn part1(allocator: std.mem.Allocator, input: []const u8) !u32 {
    var first_list = std.ArrayList(u32).init(allocator);
    defer first_list.deinit();

    var second_list = std.ArrayList(u32).init(allocator);
    defer second_list.deinit();

    var lines = std.mem.tokenizeScalar(u8, input, '\n');
    while (lines.next()) |line| {
        var it = std.mem.tokenizeScalar(u8, line, ' ');

        const first_number = try std.fmt.parseUnsigned(u32, it.next().?, 10);
        const second_number = try std.fmt.parseUnsigned(u32, it.next().?, 10);

        try first_list.append(first_number);
        try second_list.append(second_number);
    }

    std.mem.sort(u32, first_list.items, {}, comptime std.sort.asc(u32));
    std.mem.sort(u32, second_list.items, {}, comptime std.sort.asc(u32));

    var result: u32 = 0;
    for (first_list.items, second_list.items) |first, second| {
        result += if (first > second) first - second else second - first;
    }

    return result;
}

fn part2(allocator: std.mem.Allocator, input: []const u8) !u32 {
    var first_list = std.ArrayList(u32).init(allocator);
    defer first_list.deinit();

    var occurences = std.AutoHashMap(u32, u32).init(allocator);
    defer occurences.deinit();

    var lines = std.mem.tokenizeScalar(u8, input, '\n');
    while (lines.next()) |line| {
        var it = std.mem.tokenizeScalar(u8, line, ' ');

        const first_number = try std.fmt.parseUnsigned(u32, it.next().?, 10);
        const second_number = try std.fmt.parseUnsigned(u32, it.next().?, 10);

        try first_list.append(first_number);

        const entry = try occurences.getOrPut(second_number);
        if (!entry.found_existing) {
            entry.value_ptr.* = 1;
        } else {
            entry.value_ptr.* += 1;
        }
    }

    var result: u32 = 0;
    for (first_list.items) |first| {
        const count = occurences.get(first);
        if (count != null) {
            result += first * count.?;
        }
    }

    return result;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();
    defer _ = gpa.deinit();

    const file: []const u8 = @embedFile("./day-01.txt");

    const first_result = try part1(allocator, file);
    print("Part 1: {}\n", .{first_result});

    const second_result = try part2(allocator, file);
    print("Part 2: {}\n", .{second_result});
}

test "part_1" {
    const input =
        \\ 3   4
        \\ 4   3
        \\ 2   5
        \\ 1   3
        \\ 3   9
        \\ 3   3
    ;

    const result = try part1(std.testing.allocator, input);
    try std.testing.expectEqual(result, 11);
}

test "part_2" {
    const input =
        \\ 3   4
        \\ 4   3
        \\ 2   5
        \\ 1   3
        \\ 3   9
        \\ 3   3
    ;

    const result = try part2(std.testing.allocator, input);
    try std.testing.expectEqual(result, 31);
}

File: 2024/ocaml/day-01/main.ml
let input = "3   4\n4   3\n2   5\n1   3\n3   9\n3   3"

let sum_pairs (first : 'a list) (second : 'a list) =
  let rec sum (one : 'a list) (two : 'a list) (acc : int) =
    match (one, two) with
    | [], [] -> acc
    | x :: first_rest, y :: second_rest ->
        sum first_rest second_rest (max x y - min x y + acc)
    | _ -> failwith "test"
  in
  sum first second 0

let split_pairs pairs =
  let rec aux pairs acc1 acc2 =
    match pairs with
    | [] -> (acc1, acc2)
    | (x, y) :: rest -> aux rest (x :: acc1) (y :: acc2)
  in
  aux pairs [] []

let parse_input input =
    input |> String.split_on_char '\n'
    |> List.map (fun line ->
          line |> String.split_on_char ' '
          |> List.filter (fun c -> compare c "" != 0)
          |> List.map int_of_string
          |> fun values ->
          match values with
          | [ first; second ] -> (first, second)
          | _ -> failwith "invalid input")

let part_1() =
  let lines = parse_input input in
  let first, second =
    split_pairs lines |> fun (first, second) ->
    (List.sort compare first, List.sort compare second)
  in
  let result = sum_pairs first second in
  print_endline (string_of_int result)

let part_2() =
  let lines = parse_input input in

  let rec sum (list: (int * int) list) acc = match list with
    | [] -> 0
    | (first, second) :: rest ->
        let count = List.length (List.filter (fun (_, entry) -> entry == first) lines) in
        sum rest acc + first * count  in

  let result = sum lines 0 in
  print_endline (string_of_int result)


let () =
  part_1();
  part_2();
  ()
